docs/design/2022-03-09-optimize-gc-for-internal-transaction.md

```diff
@@ -0,0 +1,207 @@
+
+#### calculate gc safe point
+
+Currently, the processes TiDB calculates safe point is as bellow, it is implemented in `(is *InfoSyncer) ReportMinStartTS`
+
+- Get current time stamp from PD, save it to variable `now`,  get time stamp that is 24 hours earlier than `now`, save it to variable `startTSLowerLimit`
+- Get all user client sessions, save them to slice `processlist`
+- Traverse `processlist` to get startTS from every client session and compare the `startTS` with `now` to get the minimum time stamp which is after `startTSLowerLimit` , save it as minStartTS which is gc safe point.
+
+This design add extra processes  in `(is *InfoSyncer) ReportMinStartTS` to take account of internal transactions startTS when calculates gc safe point . The new processes TiDB calculates safe point is as bellow.
+
+- Get current time stamp from PD, save it to variable `now`,  get time stamp that is earlier than `now` which is specified by system variable `tidb_gc_txn_max_wait_time`, save it to variable `startTSLowerLimit`
+- Get all user client sessions, save them to slice `processlist`
+- Get the startTS of every internal transaction run by internal session, save it to slice `InnerSessionStartTSList`
+- Get the startTS of every internal transaction run by `RunInNewTxn`, save it to map `innerTxnStartTsMap`
+- Traverse `processlist` to get startTS from every client session and compare the `startTS` with `now` to get the minimum time stamp which is after `startTSLowerLimit` , save it as `minStartTS`.
+- Traverse `InnerSessionStartTSList` to get startTS and compare with `minStartTS` to get the minimum time stamp which is after `startTSLowerLimit` , save it also as `minStartTS`.
+- Traverse `innerTxnStartTsMap` to get startTS and compare with `minStartTS` to get the minimum time stamp which is after `startTSLowerLimit` , save it also as `minStartTS` which is gc safe point.
+
+#### store and delete internal sessions
+
+TiDB gets an internal session from a session pool implemented in `(s *session) getInternalSession`, when the transaction finished,it puts the session to the pool. It stores the session to session manager when gets an internal session, delete the session from session manager when returns the session to a pool.
+
+#### store and delete internal transactions
+
+The internal transactions here are executed by function `RunInNewTxn`. It stores the transaction to  `globalInnerTxnTsBox`  at transaction begin and deletes the transaction at end.
+
+### data structure design
+
+#### SessionManager 
+
+TiDB manages client sessions by SessionManager interface. This design expands the SessionManager interface to manage internal sessions too.
+
+```go
+type SessionManager interface {
+	// Put the internal session pointer to the map in the SessionManager
+	StoreInternalSession(addr unsafe.Pointer)
```

Comment:
If so, I think it should use `interface{}` or others. 
Does the problem " the object which is cast to unsafe.Pointer may be GC" exist? @cfzjywxk @youjiali1995 @sticnarf @MyonKeminta

Reply:
It depends on how you implement the method. If the `unsafe.Pointer` still exists, it can be recognized by the GC.
The code in https://github.com/pingcap/tidb/pull/32726 looks good because the pointer is still held by the map.

But I don't really get why you need an `unsafe.Pointer` here... Isn't a simple `*session` sufficient?

Reply:
if define `internalSessions map[*session]struct{} in `Server` struct, it will produce package cycle

Reply:
How about using the similar definition like the session pool.
The `usafe.Pointer` like parameters would make the interface difficult to understand and use.